discovery/aws/ec2_test.go

```diff
@@ -0,0 +1,467 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aws
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"strings"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// "Helper" functions.
+const (
+	chars     = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	ipv6chars = "0123456789abcdef"
+)
+
+func RandIpv4() string {
+	return fmt.Sprintf("%d.%d.%d.%d", RandNumber(1, 255), RandNumber(0, 255), RandNumber(0, 255), RandNumber(0, 255))
+}
+
+func RandIpv6() string {
+	address := "2000"
+
+	for i := 0; i < 7; i++ {
+		e := make([]byte, 4)
+		for j := range e {
+			e[j] = ipv6chars[rand.Intn(len(ipv6chars))]
+		}
+		address += ":" + string(e)
+	}
+
+	return address
+}
+
+func RandNumber(minimum, maximum int) int {
+	return rand.Intn(maximum-minimum) + minimum
+}
+
+func RandString(n int) string {
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = chars[rand.Intn(len(chars))]
+	}
+
+	return string(b)
+}
+
+// Struct for test data.
+type EC2Data struct {
+	region string
+
+	azNames  []string
+	azIDs    []string
+	azToAZID map[string]string
+
+	ownerID string
+
+	instances []*ec2.Instance
+}
+
+var ec2Data EC2Data
```

Comment:
instead of using this global variable, I wonder if you could create this on every test and them pass as parameter to the mock client.

e.g.:
```
type mockEC2Client struct {
	ec2iface.EC2API
        ec2Data ec2Data
}
```

Reply:
I hope https://github.com/prometheus/prometheus/pull/14364/commits/1774d8354499c017a8e5772a5183a6fe6390af2d is what you meant.